File: python/t76/drpd/message/messages/epr_mode.py
"""
EPR (Extended Power Range) Mode Message
"""
from enum import IntEnum
from typing import Dict

from ._base import StandardMessage


class EPRModeAction(IntEnum):
    """Action values from USB-PD 3.2 Table 6.50."""

    ENTER = 0x01
    ENTER_ACKNOWLEDGED = 0x02
    ENTER_SUCCEEDED = 0x03
    ENTER_FAILED = 0x04
    EXIT = 0x05


class EPRModeEnterFailedReason(IntEnum):
    """Enter Failed Data field values from USB-PD 3.2 Table 6.50."""

    UNKNOWN = 0x00
    CABLE_NOT_EPR_CAPABLE = 0x01
    SOURCE_FAILED_BECOME_VCONN_SOURCE = 0x02
    EPR_CAPABLE_RDO_BIT_NOT_SET = 0x03
    SOURCE_UNABLE_TO_ENTER_EPR = 0x04
    EPR_CAPABLE_PDO_BIT_NOT_SET = 0x05


class EPRModeMessage(StandardMessage):
    """
    EPR_Mode message wrapper with EPRMDO decoding (USB-PD 3.2 Table 6.50).
    - B31..24: Action
    - B23..16: Data
    - B15..0: Reserved
    """
    @property
    def name(self) -> str:
        return "EPR_Mode"

    @classmethod
    def encode(cls, action: int, data: int = 0) -> bytes:
        """
        Create a 4-byte EPRMDO.

        Args:
            action: EPR_Mode action (B31..24)
            data: EPR_Mode data (B23..16)

        Returns:
            bytes: The encoded message body (4 bytes)
        """
        eprmdo = ((action & 0xFF) << 24) | ((data & 0xFF) << 16)
        return eprmdo.to_bytes(4, "little")

    @property
    def eprmdo(self) -> int:
        """Return the raw EPRMDO word or zero if payload is too short."""
        if len(self.body) < 4:
            return 0
        return int.from_bytes(self.body[0:4], byteorder="little")

    @property
    def action(self) -> int:
        """Action field (B31..24)."""
        return (self.eprmdo >> 24) & 0xFF

    @property
    def data(self) -> int:
        """Data field (B23..16)."""
        return (self.eprmdo >> 16) & 0xFF

    @property
    def reserved(self) -> int:
        """Reserved field (B15..0)."""
        return self.eprmdo & 0xFFFF

    @property
    def action_text(self) -> str:
        """Human-readable action text."""
        mapping = {
            EPRModeAction.ENTER: "Enter",
            EPRModeAction.ENTER_ACKNOWLEDGED: "Enter Acknowledged",
            EPRModeAction.ENTER_SUCCEEDED: "Enter Succeeded",
            EPRModeAction.ENTER_FAILED: "Enter Failed",
            EPRModeAction.EXIT: "Exit",
        }
        try:
            return mapping[EPRModeAction(self.action)]
        except ValueError:
            return "Reserved/Unknown"

    @property
    def data_text(self) -> str:
        """Human-readable data text based on Action."""
        if self.action == EPRModeAction.ENTER:
            return f"Sink Operational PDP = {self.data}"
        if self.action in (
            EPRModeAction.ENTER_ACKNOWLEDGED,
            EPRModeAction.ENTER_SUCCEEDED,
            EPRModeAction.EXIT,
        ):
            return "0x00 (required)"
        if self.action == EPRModeAction.ENTER_FAILED:
            mapping = {
                EPRModeEnterFailedReason.UNKNOWN: "Unknown cause",
                EPRModeEnterFailedReason.CABLE_NOT_EPR_CAPABLE:
                    "Cable not EPR Capable",
                EPRModeEnterFailedReason.SOURCE_FAILED_BECOME_VCONN_SOURCE:
                    "Source failed to become VCONN Source",
                EPRModeEnterFailedReason.EPR_CAPABLE_RDO_BIT_NOT_SET:
                    "EPR Capable bit not set in RDO",
                EPRModeEnterFailedReason.SOURCE_UNABLE_TO_ENTER_EPR:
                    "Source unable to enter EPR Mode",
                EPRModeEnterFailedReason.EPR_CAPABLE_PDO_BIT_NOT_SET:
                    "EPR Capable bit not set in PDO",
            }
            try:
                reason = mapping[EPRModeEnterFailedReason(self.data)]
            except ValueError:
                reason = "Reserved/Unknown"
            return reason
        return "Reserved/Unknown"

    @property
    def renderable_properties(self) -> Dict[str, str]:
        """Returns a dictionary of properties for display"""
        properties = super().renderable_properties

        properties.update({
            "Action": self.action_text,
            "Details": self.data_text,
        })

        return properties

File: firmware/lib/proto/pd_messages/epr_mode.cpp
/**
 * @file epr_mode.cpp
 * @copyright Copyright (c) 2025 MTA, Inc.
 */

#include "epr_mode.hpp"
#include "../pd_message_types.hpp"
#include <cstdio>


using namespace T76::DRPD::Proto;


EPRMode::EPRMode(uint32_t raw) : _raw(raw) {
    _messageInvalid = !_validate();
}


EPRMode::EPRMode(Action action, uint8_t data) {
    // Bits 31:24 = Action
    // Bits 23:16 = Data
    // Bits 15:0 = Reserved (0)
    _raw = (static_cast<uint32_t>(action) << 24) | (static_cast<uint32_t>(data) << 16);
    _messageInvalid = !_validate();
}


std::span<const uint8_t> EPRMode::raw() const {
    // Convert 32-bit value to little-endian bytes
    _rawBytes[0] = _raw & 0xFF;
    _rawBytes[1] = (_raw >> 8) & 0xFF;
    _rawBytes[2] = (_raw >> 16) & 0xFF;
    _rawBytes[3] = (_raw >> 24) & 0xFF;

    return _rawBytes;
}


uint32_t EPRMode::numDataObjects() const {
    return 1;
}


uint32_t EPRMode::rawMessageType() const {
    return static_cast<uint32_t>(DataMessageType::EPR_Mode);
}


bool EPRMode::isMessageInvalid() const {
    return _messageInvalid;
}


EPRMode::Action EPRMode::action() const {
    return static_cast<Action>((_raw >> 24) & 0xFF);
}


void EPRMode::action(Action value) {
    _raw = (_raw & 0x00FFFFFF) | (static_cast<uint32_t>(value) << 24);
    _messageInvalid = !_validate();
}


uint8_t EPRMode::data() const {
    return (_raw >> 16) & 0xFF;
}


void EPRMode::data(uint8_t value) {
    _raw = (_raw & 0xFF00FFFF) | (static_cast<uint32_t>(value) << 16);
    _messageInvalid = !_validate();
}


bool EPRMode::_validate() {
    // Validate Action field
    uint8_t action_val = (_raw >> 24) & 0xFF;
    switch (action_val) {
        case 0x01:  // Enter
        case 0x02:  // Enter Acknowledged
        case 0x03:  // Enter Succeeded
        case 0x04:  // Enter Failed
        case 0x05:  // Exit
            break;  // Valid
        default:
            return false;  // Invalid action
    }

    // Validate that reserved bits (15:0) are zero
    if ((_raw & 0x0000FFFF) != 0) {
        return false;
    }

    return true;
}


std::string EPRMode::_actionToString(Action action) {
    switch (action) {
        case Action::Enter:
            return "Enter";
        case Action::EnterAcknowledged:
            return "Enter Acknowledged";
        case Action::EnterSucceeded:
            return "Enter Succeeded";
        case Action::EnterFailed:
            return "Enter Failed";
        case Action::Exit:
            return "Exit";
        default:
            return "Unknown";
    }
}


std::string EPRMode::_failureReasonToString(FailureReason reason) {
    switch (reason) {
        case FailureReason::UnknownCause:
            return "Unknown cause";
        case FailureReason::CableNotEprCapable:
            return "Cable not EPR Capable";
        case FailureReason::SourceNotVconnSource:
            return "Source failed to become VCONN Source";
        case FailureReason::EprCapableNotInRdo:
            return "EPR Capable bit not set in RDO";
        case FailureReason::SourceCannotEnterEpr:
            return "Source unable to enter EPR Mode";
        case FailureReason::EprCapableNotInPdo:
            return "EPR Capable bit not set in PDO";
        default:
            return "Unknown reason";
    }
}


std::string EPRMode::toString() const {
    std::string out;
    char buffer[256];

    Action act = action();
    uint8_t data_val = data();

    std::string data_str;
    if (act == Action::Enter) {
        // Data is EPR Sink Operational PDP
        data_str = "EPR Sink Operational PDP = " + std::to_string(data_val) + " PDP units";
    } else if (act == Action::EnterFailed) {
        // Data is failure reason
        data_str = "Failure Reason: " + _failureReasonToString(static_cast<FailureReason>(data_val));
    } else if (act == Action::EnterAcknowledged || act == Action::EnterSucceeded || act == Action::Exit) {
        // Data should be zero for these
        data_str = "0x" + std::to_string(data_val);
    }

    int written = std::snprintf(
        buffer,
        sizeof(buffer),
        "EPRMode\n"
        "  Raw: 0x%08X\n"
        "  Action: %s\n"
        "  Data: %s\n"
        "  Invalid: %s\n",
        _raw,
        _actionToString(act).c_str(),
        data_str.c_str(),
        isMessageInvalid() ? "yes" : "no"
    );

    if (written > 0) {
        size_t count = static_cast<size_t>(written);
        if (count >= sizeof(buffer)) {
            count = sizeof(buffer) - 1;
        }
        out.append(buffer, count);
    }

    return out;
}
